pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -199,13 +199,21 @@ func newPodInfoNoTimestamp(pod *v1.Pod) *framework.PodInfo {
 
 // activeQComp is the function used by the activeQ heap algorithm to sort pods.
 // It sorts pods based on their priority. When priorities are equal, it uses
+// Pod.CreationTimestamp. When Pod.CreationTimestamps are equal, it uses
 // PodInfo.timestamp.
 func activeQComp(podInfo1, podInfo2 interface{}) bool {
 	pInfo1 := podInfo1.(*framework.PodInfo)
 	pInfo2 := podInfo2.(*framework.PodInfo)
 	prio1 := pod.GetPodPriority(pInfo1.Pod)
 	prio2 := pod.GetPodPriority(pInfo2.Pod)
-	return (prio1 > prio2) || (prio1 == prio2 && pInfo1.Timestamp.Before(pInfo2.Timestamp))
+
+	// To make PriorityQueue near FIFO for same priority Pods, i.e. <5% Pods cut
+	// in line, we should use Pod creation time instead of the last queued time by
+	// scheduler to sort Pod for the same priority.
+	creationTime1 := pInfo1.Pod.CreationTimestamp.Time
+	creationTime2 := pInfo2.Pod.CreationTimestamp.Time
+	return (prio1 > prio2) || (prio1 == prio2 && creationTime1.Before(creationTime2)) ||
```

Comment:
This is a not a good idea. The existing logic of sorting based on last schedule attempt timestamp is fair. If we sorted pods with the same priority by their creation time, an unschedulable pod created earlier than others would get more scheduling cycles than other pods with the same priority.

Reply:
Thanks for the review :)

> If we sorted pods with the same priority by their creation time, an unschedulable pod created earlier than others would get more scheduling cycles than other pods with the same priority.

I knew this, and this is the **key goal** of this PR.

To judge whether this PR is a good idea, i.e. fair, **we need to ask our end users** of the scheduler: "if it is fair for them".

For example, for a cluster whose capacity only can serves 2 running pods at a time. UserA's JobA has 3 concurrent pods to run to complete, and each pod need to run **2hrs**, UserB's Job B has the same spec with JobA but submitted **1hrs** later, UserC's JobC has the same spec with JobB but submitted 1hrs later,..., until JobZ.

So, current scheduler logic gives nearly the same scheduling cycles for JobA, JobB...JobZ. JobA is likely to wait until JobB even JobZ to finish. So, UserA will certainly feel that current scheduler logic is very **unfair** for him, because **people in our society already built a consensus that first coming first serving, they generally cannot tolerate other people cut in line.**

See more in:
https://en.wikipedia.org/wiki/Cutting_in_line
> The act, which may be **taboo** in some instances, stands in stark contrast to the **normal** policy of first come, first served that governs **most** queue areas.

And what I mentioned in https://github.com/kubernetes/kubernetes/issues/83834:

> By default, user Pod does not have priority (i.e. 0), so usually, they expect **FIFO** scheduling in this case, even if for the PriorityQueue.
> In this requirement, the **FI(First In)** should be determined by the [Pod CreationTimestamp instead of last queued time](https://github.com/kubernetes/kubernetes/issues/83826), i.e. the **user expect longest Pending Pod since creation should be scheduled first in most cases**.
> However, we found that the PriorityQueue is far from FIFO, i.e. >40% Pods cut in line, see **How to reproduce it** for more details.


And what I mentioned in https://github.com/kubernetes/kubernetes/issues/83826:

> And even if the scheduler is not re-informed by restarting etc, using last queued time still causes https://github.com/kubernetes/kubernetes/issues/83834. And the reason is that after a Pod is poped for scheduling and then added back to queue again, its **last queued time** is updated to now (above `newPodInfo `is called again). So, if still sort by the time, it is no longer represent the **Pod pending time since created**, it just represent the **Pod pending time since last scheduled** which **user does not aware and does not care**, they just know the former Pod pending time.

Reply:
After the scheduling framework is ready, you could customize the scheduler behavior with the queue-sort extension point. But I think it would be better for us to maintain the current behavior which we used for quite a while.

Reply:
@draveness Thanks for this info, great extension point!
Some questions related:
1. Can we make the "near FIFO (sort by creation time)" as a **built-in** policy, since it is a too common requirement? So we just need to tune config to make it work in clould, such as AKS.

2. I know current behavior is used for quite a while. But it **should not be the reason to block us to rethink** "is it reasonable, i.e. fair" for our users? As we switch the default queue from FIFOQueue to PriorityQueue, we can also switch the default queuing policy for PriorityQueue to be "near FIFO (sort by creation time)", if it is more reasonable. Anyway, switch to more reasonable policy is better than insist the unreasonable one forever.

Thanks

Reply:
> I know current behavior is used for quite a while. But it should not be the reason to block us to rethink "is it reasonable, i.e. fair" for our users?

Agreed, we could have the discussion at https://github.com/kubernetes/kubernetes/issues/83834. And I'd like to hear what Bobby thinks of this. Why did we make the decisions to sort pod by last queue time instead of creation timestamp? :)

Reply:
It is already a FIFO for the pods that have never been tried before (and have the same priority). The first time "queue time" is creation time, but in subsequent retries, we use the time that a pod is added to the queue. This is roughly the same as the last scheduler retry. Why should a pod that is created earlier go to the head of the queue in all the retries? This would be unfair to other pods with the same priority.

Imagine this scenario, pods are created in this order:

queue: 1, 2, 3, 4, 5
If we implemented the algorithm proposed in this PR, the scheduler picks 1 and tries. It is unschedulable. It picks 2 and puts 1 back. The queue will look:

queue: 1, 3, 4, 5

Now, let's say 2 is unschdulable. It puts 2 back and picks up 1 again, because it is at the head of the queue. The queue will be:

queue: 2, 3, 4, 5.

And the same sequence will happen again. Basically, 3, 4, 5 will never even get a chance to be scheduled.

The existing queue algorithm puts 1 at the end of the queue after it is tried. So, other pods will get a chance to be scheduled.

Reply:
@bsalamat Thanks for this detailed reply :)
**But for your scenario, I would say 3, 4, 5 will still get chances to be scheduled.**
And this is verified in https://github.com/kubernetes/kubernetes/issues/83834, so you can see even after this fix, there are still some cut in line rate (1~10%).
So, after berifly checked our code, the true story seems like below, could you please double check it?
It mainly benefits from the unschedulable queue (and the backoff queue can also help some):
https://github.com/kubernetes/kubernetes/blob/f7091992c04e2d4f16f6b462d4a711e5b9eb0f22/pkg/scheduler/factory.go#L667-L679

So even after this PR, your scenario seems still behavior like:
**Init:**
active queue: 1, 2, 3, 4, 5

**1 failed and scheduling 2:**
active queue: 3, 4, 5
unschedulable queue: 1

**2 failed and scheduling 3:**
active queue: 4, 5
unschedulable queue: 1, 2

**3 failed and scheduling 4:**
active queue: 5
unschedulable queue: 1, 2, 3
etc.

So 3, 4, 5 will still get chances to be scheduled as tested (if there are not too frequent events in the cluster to move the unschedulable pods all the time immediately to active queue).
And this is what I have said to other approved reviewers before:

> You can think it is like the same thing as "high priority pod will not starve low priority pod" in current k8s scheduling, mentioned in wiki:
> https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#effect-of-pod-priority-on-scheduling-order
> > As a result, the higher priority Pod may be scheduled sooner than Pods with lower priority if its scheduling requirements are met. If such Pod cannot be scheduled, scheduler will continue and tries to schedule other lower priority Pods.
> 

> The orignal UT makes **extreme corner cases** that failed old Pods always immediately requeued (without the fact that we can benefit from our great features: backoff and unscheduled queues). It is not realistic, so I removed it.
> 
> In another word, this PR **sacrifices the toleration for extreme case**, i.e. too frequent events, to **achieve the near FIFO scheduling behaiour which user really get benefits in most cases.**
> 


**And let's go back to the original interesting topic. I knew that:**
> It is already a FIFO for the pods that have never been tried before (and have the same priority).

But the PR is to make it **also FIFO for the tried pods**. This is especially important for a resource tight cluster, i.e. the cluster has pending pods in most of the time.
You can see others also meet the similar issue: https://github.com/kubernetes/kubernetes/issues/83834#issuecomment-542684062

Another simple example is that:
Given 2 same pending pods A and B, but A is submitted 1 year ago, B is submitted 1 hour ago. Currently, A and B both failed to schedule, how would you plan to prioritize them in future?
If we plan to schedule B before A, what is the benefit?
Why not schedule A before B to align users' expectation?

Reply:
@yqwang-ms the scenario that I described does not happen with 5 pods because the scheduler has a default 10 second backof mechanism. If you have a large number of unschedulable pods, the scenario that I described will happen. Regardless, that example shows why not sorting by creation time is fair.

Reply:
But I tested 400 unscheduled pods that will not starve newer pods. Is it enough?

And regardless this specific PR, we can see we still have large space to improve the queuing policy fairness for users. Can we leverage unscheduled queue to be more fair? Or do you already think that current existing policy is fair enough？

Reply:
BTW, this PR is most helpful when there are many unscheduled pods of the same priority, 1) gives them an **order/subpriority** that align with pod pending time. But under the same case, 2) you say it may starve newer pods.

I think 1) and 2) are 2 different questions, if we reach an agreement on 1), we can forcus on fix 2), instead of close both 1) and 2) immediately.

For 2), a preliminary idea is that, we can improve this PR to make sure that every Pods will be scheduled in a user specified timespan.

Reply:
@yqwang-ms   If there are a large number of unschedulable  pods created long time ago in the cluster, the new pod will wait for a long time.  I think It's unfair that the old unschedulable  pod tries to schedule many times, but the new pod of the same priority  keeps waiting.

Reply:
@denkensk please tested it. In my test, 400 unschedulable pods created long time ago in the cluster, the new pod will be scheduled to a node within 10s. 

The unschedulable pods are already **determined** to be failed, so they are **suppressed** within the **unschedulable queue**, so newer pods will still get large chances to schedule (to be the head of active queue, since old pods are all **suppressed**, and some in backing off).